### PR TITLE
Add check to verify if webAuthUser is email address

### PR DESCRIPTION
**Pull request by prajith-nair**
Description: There are scenarios where application accepts X-WEBAUTH-USER as email address. We need to get login name from the email, if REVERSE_PROXY_AUTHENTICATION_USER is empty. 

Issue : https://github.com/go-gitea/gitea/issues/20433

### CODE REVIEW

File: services/auth/reverseproxy.go [hunk 1]
```diff
@@ -39,7 +39,12 @@ type ReverseProxy struct{}
 func (r *ReverseProxy) getUserName(req *http.Request) string {
 	webAuthUser := strings.TrimSpace(req.Header.Get(setting.ReverseProxyAuthUser))
 	if len(webAuthUser) == 0 {
-		return ""
+		email := strings.TrimSpace(req.Header.Get(setting.ReverseProxyAuthEmail))
+		if len(email) == 0 {
+			return ""
+		}
+		webAuthUser := strings.Split(email, "@")[0]
```
Comment (lunny) [inline services/auth/reverseproxy.go:46]: If `email` doesn't contain `@`, we need some handler.

Comment (prajith-nair) [reply services/auth/reverseproxy.go:46]:
```
if !strings.Contains(email, "@") {
			return ""
		}
		webAuthUser := strings.Split(email, "@")[0]
```

@lunny we can return null if email doesn't have a `@` Let me know your thoughts.

**Comment by prajith-nair** [comment]
@lunny Please let us know if this request will be considered in 1.18.0, as its critical for our ecosystem to allow Gitea authentication based on X-WEBAUTH-USER header as email address.

File: services/auth/reverseproxy.go (hunk 1 above)
Comment (lunny) [reply services/auth/reverseproxy.go:46]: If `@` is the first char of that string?

Comment (prajith-nair) [reply services/auth/reverseproxy.go:46]: @lunny  - 

```
// getUserName extracts the username from the "setting.ReverseProxyAuthUser" header
func (r *ReverseProxy) getUserName(req *http.Request) string {
	webAuthUser := strings.TrimSpace(req.Header.Get(setting.ReverseProxyAuthUser))
	if len(webAuthUser) == 0 {
		email := strings.TrimSpace(req.Header.Get(setting.ReverseProxyAuthEmail))
		if !strings.Contains(email, "@") {
			log.Trace("Not a valid email address %v", email)
			return nil
		} else if strings.HasPrefix(email, "@") {
			log.Fatal("Invalid email address found : %v", err)
		} else {
			webAuthUser := strings.Split(email, "@")[0]
			return webAuthUser
		}
	}
	return webAuthUser
}
```
We can add something like above to check for first character.

Comment (prajith-nair) [reply services/auth/reverseproxy.go:46]: Created a PR with latest changes https://github.com/go-gitea/gitea/pull/21264

**Comment by KN4CK3R** [comment]
I think it would be better to use a different logic.

1. Extract the username header
2. Does it look like an email? Search for the user by email
3. Does it not look like an email? Search for the user by name
4. Empty username header? Extract the email header
5. No empty? Search for the user by email

**Comment by wxiaoguang** [comment]
TBH, I do not think there should be too many "implicit logic" in the auth code.

Email is email, username is username, it seems fragile to keep "guessing" the values.

**Comment by wxiaoguang** [comment]
I think this pr/issue can be closed, feel free to reopen if there are new thoughts.

**Pull request closed without merging**